### PR TITLE
enable transient storage feature

### LINKED ISSUE #4177: [evm] enable transient storage feature
### Description of new feature request
EIP-1153 transient storage is soon to be activated in Cancun upgrade. We will enable it in our EVM in upcoming (or the next afterwards) hard-fork
1. #4151 has added the necessary data structure
2. continue to enable it in EVM. Please refer to https://github.com/ethereum/go-ethereum/pull/26003 for correct implementation

### Is this a hard-fork (and height to be activated)
- [x] Yes
- Name of activation height:

### Implementation plan
- [] Phase 1 description (PR number)
- [] Phase 2 description (PR number)
- [] Phase 3 description (PR number)

### Affected modules
- [] action
- [] actpool
- [] api
- [x] blockchain
- [] blockindex
- [] blocksync
- [] chainservice
- [] config
- [] consensus
- [] db
- [] dispatcher
- [] ioctl
- [] p2p
- [] server
- [] state
- [] tools

### PR DESCRIPTION
**Pull request by millken**
# Description
EVM upgrade to v1.13.12
enable transient storage feature
add tests

Fixes #4177

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [x] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/execution/evm/evm.go [hunk 1]
```diff
@@ -446,6 +448,8 @@ func getChainConfig(g genesis.Blockchain, height uint64, id uint32, getBlockTime
 	}
 	sumatraTimestamp := (uint64)(sumatraTime.Unix())
 	chainConfig.ShanghaiTime = &sumatraTimestamp
+	//todo: enable Cancun
+	chainConfig.CancunTime = &sumatraTimestamp
```
Comment (millken) [inline action/protocol/execution/evm/evm.go:452]: Need Hardfork cancun time

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 2]
```diff
@@ -464,25 +470,40 @@ func (stateDB *StateDBAdapter) Suicide(evmAddr common.Address) bool {
...
+		return
+	}
+
+	stateDB.setTransientState(addr, key, value)
+
+}
+
+// Selfdestruct6780 implements EIP-6780
+func (stateDB *StateDBAdapter) Selfdestruct6780(evmAddr common.Address) {
+	//Todo: implement EIP-6780
```
Comment (millken) [inline action/protocol/execution/evm/evmstatedbadapter.go:495]: implemented another PR

File: action/protocol/execution/testdata-shanghai/self-destruct.json [hunk 3]
```diff
@@ -61,10 +61,10 @@
         "rawGasLimit": 1000000,
         "rawGasPrice": "0",
         "rawAccessList": [],
-        "rawExpectedGasConsumed": 10400,
+        "rawExpectedGasConsumed": 12646,
         "expectedStatus": 1,
         "hasReturnValue": true,
-        "rawReturnValue": "",
+        "rawReturnValue": "0000000000000000000000000000000000000000000000000000000000000014",
```
Comment (millken) [inline action/protocol/execution/testdata-shanghai/self-destruct.json:67]: "selfdestruct" has been deprecated. Note that, starting from the Cancun hard fork, the underlying opcode no longer deletes the code and data associated with an account and only transfers its Ether to the beneficiary

File: Makefile [hunk 4]
```diff
@@ -11,7 +11,7 @@ GOVET=$(GOCMD) vet
 GOBUILD=$(GOCMD) build
 GOINSTALL=$(GOCMD) install
 GOCLEAN=$(GOCMD) clean
-GOTEST=$(GOCMD) test
+GOTEST=GOARCH=amd64 CGO_ENABLED=1 $(GOCMD) test
```
Comment (millken) [inline Makefile:14]: fix gomonkey error in mac m1

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -25,7 +25,7 @@ jobs:
       - name: Set up Go
         uses: actions/setup-go@v3
         with:
-          go-version: 1.19.12
+          go-version: 1.21.4
```
Comment (dustinxie) [inline .github/workflows/ci.yaml:28]: is this a must? if we want to upgrade go version, should do in another PR

File: Makefile (hunk 4 above)
Comment (dustinxie) [reply Makefile:14]: can we fix the test instead of here? this would limit the testing env to amd64 only

File: action/protocol/execution/evm/evm.go [hunk 6]
```diff
@@ -225,10 +226,11 @@ func securityDeposit(ps *Params, stateDB vm.StateDB, gasLimit uint64) error {
 		return action.ErrGasLimit
 	}
 	gasConsumed := new(big.Int).Mul(new(big.Int).SetUint64(ps.gas), ps.txCtx.GasPrice)
-	if stateDB.GetBalance(ps.txCtx.Origin).Cmp(gasConsumed) < 0 {
+	gasConsumedUint256 := uint256.NewInt(gasConsumed.Uint64())
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm.go:229]: `gasConsumed := uint256.MustFromBig(new(big.Int).Mul(new(big.Int).SetUint64(ps.gas), ps.txCtx.GasPrice))` as above L203
Uint64() may overflow right?

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 7]
```diff
@@ -42,24 +43,27 @@ type (
...
+		contractSnapshot     map[int]contractMap   // snapshots of contracts
+		suicided             deleteAccount         // account/contract calling Suicide
+		suicideSnapshot      map[int]deleteAccount // snapshots of suicide accounts
+		preimages            preimageMap
+		preimageSnapshot     map[int]preimageMap
+		accessList           *accessList // per-transaction access list
+		accessListSnapshot   map[int]*accessList
+		// Transient storage
+		transientStorage           transientStorage
+		transientStorageSnapshot   map[int]transientStorage
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:66]: this PR has lots of change, and focus on making interface work with Cancun.
let's enable transient storage in another PR

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 8]
```diff
@@ -234,8 +240,8 @@ func (stateDB *StateDBAdapter) CreateAccount(evmAddr common.Address) {
 }
 
 // SubBalance subtracts balance from account
-func (stateDB *StateDBAdapter) SubBalance(evmAddr common.Address, amount *big.Int) {
-	if amount.Cmp(big.NewInt(int64(0))) == 0 {
+func (stateDB *StateDBAdapter) SubBalance(evmAddr common.Address, amount *uint256.Int) {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:243]:
```
SubBalance(evmAddr common.Address, a256 *uint256.Int) {
  amount := a256.ToBig()
```
then no need to change inside the func

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 9]
```diff
@@ -263,9 +269,9 @@ func (stateDB *StateDBAdapter) SubBalance(evmAddr common.Address, amount *big.In
 }
 
 // AddBalance adds balance to account
-func (stateDB *StateDBAdapter) AddBalance(evmAddr common.Address, amount *big.Int) {
+func (stateDB *StateDBAdapter) AddBalance(evmAddr common.Address, amount *uint256.Int) {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:272]: same as above

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 10]
```diff
@@ -417,28 +423,28 @@ func (stateDB *StateDBAdapter) GetRefund() uint64 {
...
 	}
 	s, err := stateDB.accountState(evmAddr)
 	if err != nil {
 		log.L().Debug("Failed to get account.", zap.String("address", evmAddr.Hex()))
-		return false
+		return
 	}
 	// clears the account balance
-	actBalance := new(big.Int).Set(s.Balance)
+	actBalance := new(uint256.Int).SetBytes(s.Balance.Bytes())
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:438]: we don't need to change this?

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 11]
```diff
@@ -447,13 +453,13 @@ func (stateDB *StateDBAdapter) Suicide(evmAddr common.Address) bool {
 		// before calling Suicide, EVM will transfer the contract's balance to beneficiary
 		// need to create a transaction log on successful suicide
 		if stateDB.lastAddBalanceAmount.Cmp(actBalance) == 0 {
-			if stateDB.lastAddBalanceAmount.Cmp(big.NewInt(0)) > 0 {
+			if stateDB.lastAddBalanceAmount.Cmp(common.U2560) > 0 {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:456]: no need to change?

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 12]
```diff
@@ -447,13 +453,13 @@ func (stateDB *StateDBAdapter) Suicide(evmAddr common.Address) bool {
...
 				from, _ := address.FromBytes(evmAddr[:])
 				stateDB.addTransactionLogs(&action.TransactionLog{
 					Type:      iotextypes.TransactionLogType_IN_CONTRACT_TRANSFER,
 					Sender:    from.String(),
 					Recipient: stateDB.lastAddBalanceAddr,
-					Amount:    stateDB.lastAddBalanceAmount,
+					Amount:    stateDB.lastAddBalanceAmount.ToBig(),
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:462]: no need to change?

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 13]
```diff
@@ -464,25 +470,40 @@ func (stateDB *StateDBAdapter) Suicide(evmAddr common.Address) bool {
...
+func (stateDB *StateDBAdapter) HasSelfDestructed(evmAddr common.Address) bool {
 	addrHash := hash.BytesToHash160(evmAddr.Bytes())
 	_, ok := stateDB.suicided[addrHash]
 	return ok
 }
 
 // SetTransientState sets transient storage for a given account
 func (stateDB *StateDBAdapter) SetTransientState(addr common.Address, key, value common.Hash) {
-	log.S().Panic("SetTransientState not implemented")
+	prev := stateDB.GetTransientState(addr, key)
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:484]: let's do in next PR

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 14]
```diff
@@ -522,6 +543,8 @@ func (stateDB *StateDBAdapter) Prepare(rules params.Rules, sender, coinbase comm
 	if !rules.IsBerlin {
 		return
 	}
+	// Clear out any leftover from previous executions
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:546]: in next PR

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 15]
```diff
@@ -539,6 +562,9 @@ func (stateDB *StateDBAdapter) Prepare(rules params.Rules, sender, coinbase comm
 	if rules.IsShanghai { // EIP-3651: warm coinbase
 		stateDB.AddAddressToAccessList(coinbase)
 	}
+	// Reset transient storage at the beginning of transaction execution
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:565]: in next PR

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 16]
```diff
@@ -584,6 +610,18 @@ func (stateDB *StateDBAdapter) RevertToSnapshot(snapshot int) {
...
+	{
+		delete(stateDB.transientStorageSnapshot, snapshot)
+		for i := snapshot + 1; ; i++ {
+			if _, ok := stateDB.transientStorageSnapshot[i]; ok {
+				delete(stateDB.transientStorageSnapshot, i)
+			} else {
+				break
+			}
+		}
+	}
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:624]: in next PR

File: action/protocol/execution/evm/evmstatedbadapter_test.go [hunk 17]
```diff
@@ -933,3 +933,45 @@ func TestSortMap(t *testing.T) {
 		require.True(testFunc(t, sm))
 	})
 }
+
+func TestStateDBTransientStorage(t *testing.T) {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter_test.go:937]: in next PR

File: action/protocol/execution/testdata-shanghai/self-destruct.json (hunk 3 above)
Comment (dustinxie) [reply action/protocol/execution/testdata-shanghai/self-destruct.json:67]: this is test for Shanghai EVM, your change should not break the test

File: e2etest/contract_staking_test.go [hunk 18]
```diff
@@ -2140,8 +2140,8 @@ func stake(lsdABI abi.ABI, bc blockchain.Blockchain, sf factory.Factory, dao blo
 	r.EqualValues(iotextypes.ReceiptStatus_Success, receipts[0].Status)
 	buckets, err := indexer.Buckets(blk.Height())
 	r.NoError(err)
-	slices.SortFunc(buckets, func(i, j *contractstaking.Bucket) bool {
-		return i.Index < j.Index
+	slices.SortFunc(buckets, func(i, j *contractstaking.Bucket) int {
```
Comment (dustinxie) [inline e2etest/contract_staking_test.go:2143]: why need to change these?

File: action/protocol/execution/evm/evm.go (hunk 1 above)
Comment (dustinxie) [reply action/protocol/execution/evm/evm.go:452]: use `ToBeEnabledHeight`

File: action/protocol/execution/evm/evm_test.go [hunk 19]
```diff
@@ -347,8 +347,8 @@ func TestConstantinople(t *testing.T) {
 		require.Equal(isSumatra, chainRules.IsShanghai)
 
 		// Cancun, Prague not yet enabled
-		require.False(evmChainConfig.IsCancun(evm.Context.Time))
-		require.False(evmChainConfig.IsPrague(evm.Context.Time))
+		require.False(evmChainConfig.IsCancun(big.NewInt(0), evm.Context.Time))
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm_test.go:350]: what is the first input?

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (millken) [reply .github/workflows/ci.yaml:28]: Yes, go-ethereum has been upgraded to go1.21, we must upgrade too
